### PR TITLE
Prototype of creating a donation/periodic donation

### PR DESCRIPTION
**Pull request by nickhsine**
@babygoat

### CODE REVIEW

File: doc/donation.apib [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+### PayByCardTokenRequest
++ card_key : 91c8c5bd22208027093e7a2a7ed8cb08c616183f489088c9f1dd0ef51975efa4 (required)
++ card_token : 41ce67b0c2e6dbea2725dfef37bd71695da7405f1e7c904469d3678f191f31bc (required)
++ periodic_donation_id : 10 (required, number)
++ amount : 500 (required, number)
++ currency : TWD (optional)                         # default is TWD
++ details : 報導者小額捐款 (optional)
++ `order_number` : `order_number_we_define` (optional)
++ `merchant_id` : `twreporter_CTBC` (optional)
+
```
Comment (babygoat) [inline doc/donation.apib]: Do we need instalment?

File: doc/donation.apib [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+### CreatePeriodicDonationResponse
++ donation_result (PayByPrimeResponse, required)
++ periodic_donation_result (PeriodicPayResponse, required)
+
+### PayByPrimeFailResponse
++ prime : Required. Type: <string>. EX: test_3a2fb2b7e892b914a03c95dd4dd5dc7970c908df67a49527c0a648b2bc9 (optional)
++ amount : Required. Type: <number>. Condition: bigger than 0. EX: 500 (optional)
++ cardholder (CardholderFailData, optional)
++ `pay_by` : Required. Type: <string>. EX: credit_card (optional)
+
```
Comment (babygoat) [inline doc/donation.apib]: Should we deal with the error detected by Tap Pay Fraud Detector( missing `details`, `credit_card_data`)?

File: doc/donation.apib [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
++ card_info (CardInfo, required)
++ `order_number` : `order_number_we_define` (required)
++ acquirer : TW_CTBC (required)
++ transaction_time_millis : 1531363227359 (required, number)
++ bank_transaction_time (BankTransactionTime, required)
++ bank_result_code : 0 (required)
++ bank_result_msg : bank_message (required)
+
+### Cardholder
++ phone_number : +886912345678 (required)
```
Comment (babygoat) [inline doc/donation.apib]: It seemed that the E.164 SHOULD be enforced.  What's the plan for the frontend input field?

File: doc/donation.apib [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            "message": "Unknown server error during create"
+            }
+
+# View Donations [/v1/donations{?email,pay_methods,page,number_per_page}]
+## View Donations of User by Email [GET]
++ Parameters
+    + email: `developer%40twreporter.org` (required) - User's email.
+    + page: `1` (optional, unsigned integer) - The page number, default to 1 if not specified.
+    + `pay_methods`: `credit_card` (optional) - A collection of pay methods.
+    + `number_per_page`: `5` (optional, unsigned integer) - The number of notification per page, default to 20 if not specified.
```
Comment (babygoat) [inline doc/donation.apib]: number_per_page  or  n_per_page as states in response?

File: doc/donation.apib [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            "status": "error",
+            "message": "Unknown server error during create"
+            }
+
+# View Donations [/v1/donations{?email,pay_methods,page,number_per_page}]
+## View Donations of User by Email [GET]
++ Parameters
+    + email: `developer%40twreporter.org` (required) - User's email.
+    + page: `1` (optional, unsigned integer) - The page number, default to 1 if not specified.
+    + `pay_methods`: `credit_card` (optional) - A collection of pay methods.
```
Comment (babygoat) [inline doc/donation.apib]: pay_method?

File: doc/donation.apib [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+
+### CreatePeriodicDonationResponse
++ donation_result (PaySuccessResponse, required)
+
+### PayByPrimeFailResponse
++ prime : Required. Type: <string>. EX: test_3a2fb2b7e892b914a03c95dd4dd5dc7970c908df67a49527c0a648b2bc9 (optional)
++ amount : Required. Type: <number>. Condition: bigger than 0. EX: 500 (optional)
++ cardholder (CardholderFailData, optional)
++ `pay_by` : Required. Type: <string>. EX: credit_card (optional)
```
Comment (babygoat) [inline doc/donation.apib]: EX: -> e.g.,?

**Comment by babygoat** [comment]
I suggest we should also record the entry creation time (& update time if necessary) as well.

File: doc/donation.apib (hunk 1 above)
Comment (nickhsine) [reply doc/donation.apib]: we don't need instalment.

File: doc/donation.apib [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
++ card_info (CardInfo, required) 
++ amount : 500 (required, number)
++ currency : TWD (required)
++ details : 報導者小額捐款 (required)
++ `order_number` : `order_number_we_define` (optional)
+
+### PeriodicPayResponse
++ id : 10 (required, number)
++ cardholder (Cardholder, required)
++ start_date : `2018-06-12` (required)        # the date the user start to pay periodically
```
[the remark this answers is not in the document]
Comment (nickhsine) [reply doc/donation.apib]: changed

File: doc/donation.apib (hunk 3 above)
Comment (nickhsine) [reply doc/donation.apib]: I will validate the phone number format on the browser, but we need the validation as well on the server side.

File: doc/donation.apib (hunk 2 above)
Comment (nickhsine) [reply doc/donation.apib]: omit fraud detector for now

File: doc/donation.apib (hunk 5 above)
Comment (nickhsine) [reply doc/donation.apib]: `pay_methods`
will be like `/v1/donations?pay_methods=credit_card&pay_methods=line&pay_methods=apple`, which is an array in URL query string.

File: doc/donation.apib (hunk 4 above)
Comment (nickhsine) [reply doc/donation.apib]: will change the response field to snake_case rather than camelCase

**Comment by babygoat** [comment]
Do we need entry deletion?(i.e., require deleted_at)

**Comment by nickhsine** [comment]
Even though we don't provide refunding now, but I think we can remain `deleted_at` for future usage.

File: membership_user.sql [hunk 8]
```diff
@@ -194,3 +194,174 @@ CREATE TABLE `web_push_subs` (
...
+  `auth_code` varchar(6) NOT NULL,
+  `amount` int(10) unsigned NOT NULL,
+  `currency` char(3) DEFAULT 'TWD' NOT NULL,
+  `order_number` varchar(50) NOT NULL,
+  `acquirer` varchar(50) NOT NULL,
+  `transaction_time` timestamp NULL DEFAULT NULL,
+  `bank_transaction_start_time` timestamp NULL DEFAULT NULL,
+  `bank_transaction_end_time` timestamp NULL DEFAULT NULL,
+  `bank_result_code` varchar(50) DEFAULT '0' NOT NULL,
+  `bank_result_msg` varchar(50) DEFAULT '' NOT NULL,
```
Comment (babygoat) [inline membership_user.sql]: Why should we treat these two fields `bank_result_code` and `bank_result_msg` be `NOT NULL`?

File: membership_user.sql [hunk 9]
```diff
@@ -207,8 +207,7 @@ CREATE TABLE `pay_by_prime_donations` (
   `created_at` timestamp DEFAULT CURRENT_TIMESTAMP,
   `updated_at` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
   `deleted_at` timestamp NULL DEFAULT NULL,
-  `email` varchar(100) NOT NULL,
-  `name` varchar(30) NOT NULL,
+  `user_id` int(10) unsigned NOT NULL,
```
Comment (babygoat) [inline membership_user.sql]: remember to add foreigner key constraint

File: membership_user.sql [hunk 10]
```diff
@@ -194,3 +194,185 @@ CREATE TABLE `web_push_subs` (
...
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `created_at` timestamp DEFAULT CURRENT_TIMESTAMP,
+  `updated_at` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  `deleted_at` timestamp NULL DEFAULT NULL,
+  `email` varchar(100) NOT NULL,
+  `pay_method` varchar(50) NOT NULL,
+  `amount` int(10) unsigned NOT NULL,
+  `currency` char(3) DEFAULT 'TWD' NOT NULL,
+  `details` varchar(50) NOT NULL,
+  `merchat_id` varchar(30) NOT NULL,
```
Comment (babygoat) [inline membership_user.sql]: merchant_id? and should we need the merchant_id in other payby tables?

Comment (nickhsine) [reply membership_user.sql]: I think yes. It is related to the bank which receives the money.
If donators give their money to us through that bank, it needs to be recorded.

File: membership_user.sql [hunk 11]
```diff
@@ -194,3 +194,185 @@ CREATE TABLE `web_push_subs` (
...
+CREATE TABLE `pay_by_other_method_donations` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `created_at` timestamp DEFAULT CURRENT_TIMESTAMP,
+  `updated_at` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  `deleted_at` timestamp NULL DEFAULT NULL,
+  `email` varchar(100) NOT NULL,
+  `pay_method` varchar(50) NOT NULL,
+  `amount` int(10) unsigned NOT NULL,
+  `currency` char(3) DEFAULT 'TWD' NOT NULL,
+  `details` varchar(50) NOT NULL,
```
Comment (babygoat) [inline membership_user.sql]: do we need this field in other payby tables?

Comment (nickhsine) [reply membership_user.sql]: yes. When users email us, they may have something to say to us. We can record that in details.

File: membership_user.sql [hunk 12]
```diff
@@ -194,3 +194,185 @@ CREATE TABLE `web_push_subs` (
...
+  `card_info_last_four` varchar(4) DEFAULT NULL,
+  `card_info_issuer` varchar(50) DEFAULT NULL,
+  `card_info_funding` tinyint DEFAULT NULL,  
+  `card_info_type` tinyint DEFAULT NULL, 
+  `card_info_level` varchar(10) DEFAULT NULL, 
+  `card_info_country` varchar(30) DEFAULT NULL, 
+  `card_info_country_code` varchar(10) DEFAULT NULL, 
+  `card_info_expiry_date` varchar(6) DEFAULT NULL, 
+  PRIMARY KEY (`id`),
+  KEY `idx_pay_by_prime_donations_email_pay_method` (`email`,`pay_method`)
```
Comment (babygoat) [inline membership_user.sql]: `cardholder_email`?

File: membership_user.sql [hunk 13]
```diff
@@ -194,3 +194,185 @@ CREATE TABLE `web_push_subs` (
...
+  `card_info_last_four` varchar(4) DEFAULT NULL,
+  `card_info_issuer` varchar(50) DEFAULT NULL,
+  `card_info_funding` tinyint DEFAULT NULL,  
+  `card_info_type` tinyint DEFAULT NULL, 
+  `card_info_level` varchar(10) DEFAULT NULL, 
+  `card_info_country` varchar(30) DEFAULT NULL, 
+  `card_info_country_code` varchar(10) DEFAULT NULL, 
+  `card_info_expiry_date` varchar(6) DEFAULT NULL, 
+  PRIMARY KEY (`id`),
+  KEY `fk_periodic_donations_idx` (`user_id`),
```
Comment (babygoat) [inline membership_user.sql]: Do we need another non-unique key constraint?

**Comment by nickhsine** [comment]
All fixed due to code review

**Comment by babygoat** [comment]
summary:
Table pay_by_prime_donations
* missing `details`, `merchant_id` fields
* make `bank_result_code`, `bank_result_msg` fields to be nullable

Table pay_by_card_token_donations
* missing `details`, `merchant_id` fields
* make `bank_result_code`, `bank_result_msg` fields to be nullable

**Comment by nickhsine** [comment]
done

**Comment by nickhsine** [comment]
New commit for request authorization.

File: tests/controller_post_test.go [hunk 14]
```diff
@@ -13,7 +13,7 @@ import (
 )
 
 /*
-The whole testing mongodb is set by ./test.go
+The whole testing mongodb is set by ./db.go
 You should check #SetMgoDefaultRecords function,
```
Comment (babygoat) [inline tests/controller_post_test.go]: setMgoDefaultRecords?

File: tests/controller_post_test.go [hunk 15]
```diff
@@ -13,7 +13,7 @@ import (
 )
 
 /*
-The whole testing mongodb is set by ./test.go
+The whole testing mongodb is set by ./db.go
```
Comment (babygoat) [inline tests/controller_post_test.go]: pre_test_environment_setup.go

File: tests/structs.go [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/jinzhu/gorm"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"twreporter.org/go-api/models"
+)
+
+type DefaultVariables struct {
```
Comment (babygoat) [inline tests/structs.go]: Do we need to export this variable?

File: tests/structs.go [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+	ThemeCol models.Theme
+
+	MockPostSlug1 string
+	MockTopicSlug string
+
+	// web push
+	WebPushEndpoint string
+}
+
+type GlobalVariables struct {
```
Comment (babygoat) [inline tests/structs.go]: Do we need to export GlobalVariables and WebPushSubscriptionPostBody?

File: tests/pre_test_environment_setup.go [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+	}
+	session, err = mgo.Dial(dbhost)
+
+	// set settings
+	utils.Cfg.MongoDBSettings.DBName = mgoDBName
+
+	return
+}
+
+func SetUpDBEnvironment() (*gorm.DB, *mgo.Session) {
```
Comment (babygoat) [inline tests/pre_test_environment_setup.go]: Do we need to export this function?

File: tests/mock_data_setup.go [hunk 19]
```diff
@@ -0,0 +1,24 @@
+package tests
+
+import (
+	"encoding/json"
+)
+
+// setDefaultWebPushSubscription - set up default records in web_push_subscriptions table
```
Comment (babygoat) [inline tests/mock_data_setup.go]: web_push_subs?

File: doc/donation.apib [hunk 20]
```diff
@@ -32,6 +32,7 @@ TWReporter Donation API, integrated with Tappay Web Service
 ### PaySuccessResponse
 + id : 11 (required, number)
 + is_periodic : true (required, boolean)
++ `pay_mehtod` : `credit_card` (required)
```
Comment (babygoat) [inline doc/donation.apib]: pay_method

**Comment by nickhsine** [comment]
fixed.

File: tests/controller_donations_test.go [hunk 21]
```diff
@@ -0,0 +1,618 @@
...
+type (
+	cardInfo struct {
+		BinCode     string `json:"bin_code"`
+		LastFour    string `json:"last_four"`
+		Issuer      string `json:"issuer"`
+		Funding     int    `json:"funding"`
+		Type        int    `json:"type"`
+		Level       string `json:"level"`
+		Country     string `json:"country"`
+		CountryCode string `json:"us"`
```
Comment (babygoat) [inline tests/controller_donations_test.go]: `json:"country_code"`

File: tests/controller_donations_test.go [hunk 22]
```diff
@@ -0,0 +1,618 @@
...
+func testCardInfoWithDefaultValue(t *testing.T, ci cardInfo) {
+	assert.Equal(t, "424242", ci.BinCode)
+	assert.Equal(t, "4242", ci.LastFour)
+	assert.Equal(t, "JPMORGAN CHASE BANK NA", ci.Issuer)
+	assert.Equal(t, 0, ci.Funding)
+	assert.Equal(t, 1, ci.Type)
+	assert.Equal(t, "", ci.Level)
+	assert.Equal(t, "UNITED STATES", ci.Country)
+	assert.Equal(t, "US", ci.CountryCode)
+	assert.Equal(t, "202301", ci.ExpiryDate)
```
Comment (babygoat) [inline tests/controller_donations_test.go]: Tap pay will not response with `expiry_date` for one-time donation ( remember=false).

File: membership_user.sql [hunk 23]
```diff
@@ -194,3 +194,193 @@ CREATE TABLE `web_push_subs` (
...
+  `updated_at` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  `deleted_at` timestamp NULL DEFAULT NULL,
+  `user_id` int(10) unsigned NOT NULL,
+  `details` varchar(50) NOT NULL,
+  `merchat_id` varchar(30) NOT NULL,
+  `amount` int(10) unsigned NOT NULL,
+  `order_number` varchar(50) NOT NULL,
+  `currency` char(3) DEFAULT 'TWD' NOT NULL,
+  `pay_method` enum('credit_card', 'line', 'apple', 'google', 'samsung') NOT NULL,
+  `status` enum('to_pay', 'paying', 'paid') NOT NULL,
```
Comment (babygoat) [inline membership_user.sql]: I suggest we should differentiate `paid` (third_party_status=0) from `error` (third_party_status!=0). (i.e., add `error` enum)

File: tests/controller_donations_test.go [hunk 24]
```diff
@@ -0,0 +1,618 @@
...
+	}
+	requestBody struct {
+		Prime       string     `json:"prime"`
+		Amount      uint       `json:"amount"`
+		Currency    string     `json:"currency"`
+		Details     string     `json:"details"`
+		Cardholder  cardholder `json:"cardholder"`
+		OrderNumber string     `json:"order_number"`
+		MerchantID  string     `json:"merchat_id"`
+		ResultURL   string     `json:"result_url"` // Line pay needed only
```
Comment (babygoat) [inline tests/controller_donations_test.go]: result_url is a json object with `frontend_redirect_url` and `backend_redirect_url` fields.

File: membership_user.sql [hunk 25]
```diff
@@ -194,3 +194,193 @@ CREATE TABLE `web_push_subs` (
...
+  `bank_transaction_start_time` timestamp NULL DEFAULT NULL,
+  `bank_transaction_end_time` timestamp NULL DEFAULT NULL,
+  `bank_result_code` varchar(50) NULL DEFAULT NULL,
+  `bank_result_msg` varchar(50) NULL DEFAULT NULL,
+  `cardholder_email` varchar(100) NOT NULL,
+  `cardholder_phone_number` varchar(20) DEFAULT NULL,
+  `cardholder_name` varchar(30) DEFAULT NULL,
+  `cardholder_zip_code` varchar(10) DEFAULT NULL,
+  `cardholder_address` varchar(100) DEFAULT NULL,
+  `cardholder_nation_id` varchar(20) DEFAULT NULL,
```
Comment (babygoat) [inline membership_user.sql]: `cardholder_national_id`

File: membership_user.sql [hunk 26]
```diff
@@ -194,3 +194,193 @@ CREATE TABLE `web_push_subs` (
...
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!40101 SET character_set_client = utf8 */;
+CREATE TABLE `pay_by_prime_donations` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `created_at` timestamp DEFAULT CURRENT_TIMESTAMP,
+  `updated_at` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  `deleted_at` timestamp NULL DEFAULT NULL,
+  `user_id` int(10) unsigned NOT NULL,
+  `details` varchar(50) NOT NULL,
+  `merchat_id` varchar(30) NOT NULL,
```
Comment (babygoat) [inline membership_user.sql]: merchant_id

**Comment by babygoat** [comment]
Please re-submit the changes to `feature-donations` branch

**Comment by nickhsine** [comment]
duplicate of [141](https://github.com/twreporter/go-api/pull/141)

**Pull request closed without merging**